When ~/.ssh/config is missing then get a exception

When you don't have a ~/.ssh/config all tests using the sshadaptor fail with
```
nl.esciencecenter.xenon.XenonException: ssh adaptor: Could not set OpenSSH config file
        at nl.esciencecenter.xenon.adaptors.ssh.SshAdaptor.setConfigFile(SshAdaptor.java:196)
        at nl.esciencecenter.xenon.adaptors.ssh.SshAdaptor.<init>(SshAdaptor.java:165)
        at nl.esciencecenter.xenon.adaptors.ssh.SshAdaptor.<init>(SshAdaptor.java:142)
        at nl.esciencecenter.xenon.engine.XenonEngine.loadAdaptors(XenonEngine.java:182)
        at nl.esciencecenter.xenon.engine.XenonEngine.<init>(XenonEngine.java:169)
        at nl.esciencecenter.xenon.engine.XenonEngine.newXenon(XenonEngine.java:92)
        at nl.esciencecenter.xenon.XenonFactory.newXenon(XenonFactory.java:57)
        at nl.esciencecenter.xenon.adaptors.GenericJobAdaptorTestParent.prepare(GenericJobAdaptorTestParent.java:154)
Caused by: java.io.FileNotFoundException: /home/xenon/.ssh/config (No such file or directory)
        at java.io.FileInputStream.open(Native Method)
        at java.io.FileInputStream.<init>(FileInputStream.java:146)
        at java.io.FileInputStream.<init>(FileInputStream.java:101)
        at com.jcraft.jsch.Util.fromFile(Util.java:492)
        at com.jcraft.jsch.OpenSSHConfig.parseFile(OpenSSHConfig.java:97)
        at nl.esciencecenter.xenon.adaptors.ssh.SshAdaptor.setConfigFile(SshAdaptor.java:193)

```

I expected if ~/.ssh/config does not exist it is not read, but instead it always tries to read it and if absent throws an exception.

It is possible to disable reading of the file with the `loadSshConfig` property set to false, but I would like to read it if its there by default.

--- a/src/nl/esciencecenter/xenon/adaptors/generic/Location.java
+++ b/src/nl/esciencecenter/xenon/adaptors/generic/Location.java
@@ -26,7 +26,7 @@ public abstract class Location {
     protected Location(String location) throws InvalidLocationException {
         // Augment location with dummy scheme, if needed, to allow processing by URI
         String augmentedLocation = null;
-        boolean hasScheme = location.indexOf(SCHEME_SEPARATOR) >= 0;
+        boolean hasScheme = location.contains(SCHEME_SEPARATOR);
         if (!hasScheme) {
             augmentedLocation = DUMMY_SCHEME + SCHEME_SEPARATOR + location;
         }
@@ -35,12 +35,15 @@ public abstract class Location {
             URI url = new URI(augmentedLocation == null ? location : augmentedLocation);
             user = url.getUserInfo();
             host = url.getHost();
-            port = url.getPort();// == -1 ? getDefaultPort() : url.getPort();
+            port = url.getPort();
             scheme = hasScheme ? url.getScheme() : null;
             path = url.getPath();
             validate();
         } catch (URISyntaxException e) {
             throw new InvalidLocationException(getAdaptorName(), "Could not parse location " + location, e);
+        }
+        if (port <= 0 && port != -1) {
+            throw new InvalidLocationException(getAdaptorName(), "Port number of " + location + " must be positive or omited");
         }
     }
 
@@ -59,7 +62,7 @@ public abstract class Location {
     }
 
     public int getPort() {
-    	if (port == -1){
+    	if (port == -1) {
     		return getDefaultPort();
     	}
     
@@ -80,7 +83,7 @@ public abstract class Location {
 
     @Override
     public String toString() {
-        StringBuilder result = new StringBuilder();
+        StringBuilder result = new StringBuilder(100);
         appendScheme(result);
         appendUser(result);
         appendHostAndPort(result);

--- a/src/nl/esciencecenter/xenon/adaptors/ssh/SshAdaptor.java
+++ b/src/nl/esciencecenter/xenon/adaptors/ssh/SshAdaptor.java
@@ -34,7 +34,6 @@ import nl.esciencecenter.xenon.engine.util.ImmutableArray;
 import nl.esciencecenter.xenon.files.Files;
 import nl.esciencecenter.xenon.files.NoSuchPathException;
 import nl.esciencecenter.xenon.jobs.Jobs;
-import nl.esciencecenter.xenon.util.Utils;
 
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
@@ -162,7 +161,7 @@ public class SshAdaptor extends Adaptor {
                 sshConfig = System.getProperty("user.home") + "/.ssh/config";
             }
             LOGGER.debug("Setting ssh known hosts file to: " + sshConfig);
-            setConfigFile(sshConfig);
+            setConfigFile(sshConfig, !getProperties().propertySet(SshAdaptor.LOAD_SSH_CONFIG));
         }
     }
 
@@ -188,12 +187,16 @@ public class SshAdaptor extends Adaptor {
         }
     }
     
-    private void setConfigFile(String sshConfigFile) throws XenonException {
+    private void setConfigFile(String sshConfigFile, boolean ignoreFail) throws XenonException {
         try {
             ConfigRepository configRepository = com.jcraft.jsch.OpenSSHConfig.parseFile(sshConfigFile);
             jsch.setConfigRepository(configRepository);
         } catch (IOException ex) {
-            throw new XenonException(SshAdaptor.ADAPTOR_NAME, "Could not set OpenSSH config file", ex);
+            if (ignoreFail) {
+                LOGGER.warn("OpenSSH config file cannot be read.", ex);
+            } else {
+                throw new XenonException(SshAdaptor.ADAPTOR_NAME, "Cannot read OpenSSH config file", ex);
+            }
         }
     }
 
@@ -295,7 +298,7 @@ public class SshAdaptor extends Adaptor {
 
     @Override
     public Map<String, String> getAdaptorSpecificInformation() {
-        Map<String,String> result = Utils.emptyMap(1);
+        Map<String,String> result = new HashMap<>(2);
         jobsAdaptor.getAdaptorSpecificInformation(result);
         return result;
     }

--- a/test/src/nl/esciencecenter/xenon/adaptors/ssh/SSHLocationTest.java
+++ b/test/src/nl/esciencecenter/xenon/adaptors/ssh/SSHLocationTest.java
@@ -77,30 +77,25 @@ public class SSHLocationTest {
         assertEquals(url, tmp.toString());
     }
 
-    @Test
-    public void test_parse_userHostDefaultPort1() throws Exception {
-        SshLocation tmp = SshLocation.parse("user@host:-42", ConfigRepository.nullConfig);
-        assertEquals("user", tmp.getUser());
-        assertEquals("host", tmp.getHost());
-        assertEquals(DEFAULT_PORT, tmp.getPort());
-    }
-
-    @Test
-    public void test_parse_userHostDefaultPort2() throws Exception {
-        SshLocation tmp = SshLocation.parse("user@host:0", ConfigRepository.nullConfig);
-        assertEquals("user", tmp.getUser());
-        assertEquals("host", tmp.getHost());
-        assertEquals(DEFAULT_PORT, tmp.getPort());
+    @Test(expected = InvalidLocationException.class)
+    public void test_parse_negativePort() throws Exception {
+        SshLocation.parse("user@host:-42", ConfigRepository.nullConfig);
     }
 
     @Test(expected = InvalidLocationException.class)
+    public void test_parse_zeroPort() throws Exception {
+        SshLocation.parse("user@host:0", ConfigRepository.nullConfig);
+    }
+
+    @Test
     public void test_parse_missingUser() throws Exception {
-        SshLocation.parse("@host:33", ConfigRepository.nullConfig);
+        SshLocation tmp = SshLocation.parse("@host:33", ConfigRepository.nullConfig);
+        assertNull(tmp.getUser());
     }
 
-    @Test(expected = InvalidLocationException.class)
     public void test_parse_missingPort() throws Exception {
-        SshLocation.parse("host:", ConfigRepository.nullConfig);
+        SshLocation tmp = SshLocation.parse("host:", ConfigRepository.nullConfig);
+        assertTrue(tmp.usesDefaultPort());
     }
 
     @Test(expected = InvalidLocationException.class)

